### PR TITLE
Bytecode Root Solver Library

### PR DESCRIPTION
**Pull request by bitzoic**
## Type of change

- New feature

## Changes

The following changes have been made:

- Added `compute_bytecode_root()` function
- Added `compute_bytecode_root_with_configurables()` function
- Added `compute_predicate_address()` function
- Added `compute_predicate_address_with_configurables()` function
- Added `predicate_address_from_root()` function
- Added `swap_configurables()` function
- Added `verify_contract_bytecode()` function
- Added `verify_contract_bytecode_with_configurables()` function
- Added `verify_predicate_address()` function
- Added `verify_predicate_address_with_configurables()` function

## Notes

- While computationally expensive, these functions will allow for certain use cases such as contract and predicate factories

### CODE REVIEW

File: libs/bytecode/src/utils.sw [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+/// Takes some bytecode and computes the resulting leaves for a merkle tree.
+fn generate_leaves(bytecode: Vec<u8>) -> Vec<b256> {
+    // Number of leaves is '(bytecode.len() / LEAF_SIZE)' if it's perfectly divisible by LEAF_SIZE.
+    // Otherwise `(bytecode.len() / LEAF_SIZE) + 1` to account for padding.
+    let num_leaves = match bytecode.len() % LEAF_SIZE {
+        0 => bytecode.len() / LEAF_SIZE,
+        _ => bytecode.len() / LEAF_SIZE + 1,
+    };
+    let mut leaves_digest: Vec<b256> = Vec::new();
```
Comment (K1-R1) [inline libs/bytecode/src/utils.sw:156]: Can this be switched to `with_capacity()` using `num_leaves`?

File: libs/bytecode/src/lib.sw [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+/// fn foo(predicate_bytecode_root: b256) {
+///     let predicate_address = generate_predicate_address(predicate_bytecode_root);
+///     assert(predicate_address != Address::from(ZERO_B256));
+/// }
+/// ```
+pub fn generate_predicate_address(bytecode_root: b256) -> Address {
+    _generate_predicate_address(bytecode_root)
+}
+
+/// Swaps out configurable values in a contract or predicate's bytecode.
```
Comment (K1-R1) [inline libs/bytecode/src/lib.sw:145]:
```suggestion
/// Swaps out configurable values in a contract's or predicate's bytecode.
```

File: tests/src/bytecode/tests/functions/swap_configurables.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                    .unwrap(),
+                my_configurables.get(0).unwrap().1.get(iter).unwrap()
+            );
+        }
+
+        // Deploy the new simple contract with the bytecode that contains the changes
+        let simple_contract_instance =
+            deploy_simple_contract_from_bytecode(wallet, result_bytecode).await;
+
+        // Assert that we get the exected value for the configurable from the deployed simple contract.
```
Comment (K1-R1) [inline tests/src/bytecode/tests/functions/swap_configurables.rs:47]:
```suggestion
        // Assert that we get the expected value for the configurable from the deployed simple contract.
```

**Review by K1-R1 — COMMENTED** [review]
Very cool 🚀. Are the READMEs awaiting the approval of SRC-11?

File: libs/bytecode/src/utils.sw (hunk 1 above)
Comment (bitzoic) [reply libs/bytecode/src/utils.sw]: This has been converted into a `raw_slice` in my recent optimization commit

**Comment by bitzoic** [comment]
Dependent on https://github.com/FuelLabs/sway-libs/pull/212 being merged first

**Comment by bitzoic** [comment]
> Very cool 🚀. Are the READMEs awaiting the approval of SRC-11?

They've just been added 😅

**Review by K1-R1 — DISMISSED** [review]

**Pull request merged**